src/query/service/src/servers/http/middleware/session.rs

```diff
@@ -537,46 +539,51 @@ impl<E: Endpoint> Endpoint for HTTPSessionEndpoint<E> {
-                        )));
-                    }
-                    Ok(Some(node)) => {
-                        let local_id = GlobalConfig::instance().query.node_id.clone();
-                        if node.id != local_id {
-                            log::info!(
-                                "forwarding /v1{} from {} to warehouse {}({})",
-                                req.uri(),
-                                local_id,
-                                warehouse,
-                                node.id
-                            );
-                            return forward_request(req, node).await;
+                let resources_management = GlobalInstance::get::<Arc<dyn ResourcesManagement>>();
+                if resources_management.support_forward_warehouse_request() {
+                    req.headers_mut().remove(HEADER_WAREHOUSE);
+
+                    let warehouse = warehouse
+                        .to_str()
+                        .map_err(|e| {
+                            HttpErrorCode::bad_request(ErrorCode::BadArguments(format!(
+                                "Invalid Header ({HEADER_WAREHOUSE}: {warehouse:?}): {e}"
+                            )))
+                        })?
+                        .to_string();
+
+                    let cluster_discovery = ClusterDiscovery::instance();
+
+                    let forward_node = cluster_discovery.find_node_by_warehouse(&warehouse).await;
+
+                    match forward_node {
+                        Err(error) => {
+                            return Err(HttpErrorCode::server_error(error).into());
+                        }
+                        Ok(None) => {
+                            let msg = format!("Not find the '{}' warehouse; it is possible that all nodes of the warehouse have gone offline. Please exit the client and reconnect, or use `use warehouse <new_warehouse>`", warehouse);
+                            warn!("{}", msg);
+                            return Err(Error::from(HttpErrorCode::bad_request(
+                                ErrorCode::UnknownWarehouse(msg),
+                            )));
+                        }
+                        Ok(Some(node)) => {
+                            let local_id = GlobalConfig::instance().query.node_id.clone();
+                            if node.id != local_id {
+                                log::info!(
+                                    "forwarding /v1{} from {} to warehouse {}({})",
+                                    req.uri(),
+                                    local_id,
+                                    warehouse,
+                                    node.id
+                                );
+                                return forward_request(req, node).await;
+                            }
                         }
                     }
                 }
+
+                log::warn!("Ignore header ({HEADER_WAREHOUSE}: {warehouse:?})");
```

Comment:
why add this log？

Reply:
The `X-DATABEND-WAREHOUSE` header should not be passed in self-managed mode. However, for compatibility, we allow and ignore this case.